server/test/write.go

```diff
@@ -733,6 +752,335 @@ var writeCommandTests = []writeCommandTest{
+					},
+				},
+			},
+		},
+		request: &openfgapb.WriteRequest{
+			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("document:budget", "reader", "user:bob"),
+			}},
+		},
+	},
+	{
+		_name: "Write fails if user field contains a type that is not allowed by the authorization model (which only allows group:...#member)",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type:      "user",
+				Relations: map[string]*openfgapb.Userset{},
+			},
+			{
+				Type: "group",
+				Relations: map[string]*openfgapb.Userset{
+					"member": {Userset: &openfgapb.Userset_This{}},
+				},
+			},
+			{
+				Type: "document",
+				Relations: map[string]*openfgapb.Userset{
+					"reader": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"reader": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type:     "group",
+									Relation: "member",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		request: &openfgapb.WriteRequest{
+			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("document:budget", "reader", "user:abc"),
+			}},
+		},
+		err: serverErrors.InvalidTuple("Object of type user is not allowed to have relation reader with document:budget",
+			tuple.NewTupleKey("document:budget", "reader", "user:abc"),
+		),
+	},
+	{
+		_name: "Write succeeds if user field contains a type that is allowed by the authorization model (which only allows group:...#member)",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type:      "user",
+				Relations: map[string]*openfgapb.Userset{},
+			},
+			{
+				Type: "group",
+				Relations: map[string]*openfgapb.Userset{
+					"member": {Userset: &openfgapb.Userset_This{}},
+				},
+			},
+			{
+				Type: "document",
+				Relations: map[string]*openfgapb.Userset{
+					"reader": {Userset: &openfgapb.Userset_This{}},
+				},
+				Metadata: &openfgapb.Metadata{
+					Relations: map[string]*openfgapb.RelationMetadata{
+						"reader": {
+							DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+								{
+									Type:     "group",
+									Relation: "member",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		request: &openfgapb.WriteRequest{
+			Writes: &openfgapb.TupleKeys{TupleKeys: []*openfgapb.TupleKey{
+				tuple.NewTupleKey("document:budget", "reader", "group:abc#member"),
+			}},
+		},
+	},
+	{
+		_name: "Write succeeds if user is * and type references a specific type",
+		// state
+		schemaVersion: typesystem.SchemaVersion1_1,
+		typeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type:      "user",
+				Relations: map[string]*openfgapb.Userset{},
+			},
+			{
+				Type: "group",
+				Relations: map[string]*openfgapb.Userset{
```

Comment:
This isn't needed as we don't define any relations on it.

Reply:
🤔 sorry I don't understand this

Reply:
I just mean let's drop this, because it shouldn't be needed. If we omit the map then it will be nil, and that should be fine.

Reply:
Ahh okay. I was confused because I thought this should fail validation, but i realized i am not calling `commands.WriteAuthorizationModel` but `storage.WriteAuthorizationModel`